### PR TITLE
CATTY-263 Use UserList in VariablesContainer and model classes

### PR DESCRIPTION
**Pull request by DhrubojyotiBis1**
- Changed the **interface** of methods in the **VariablesContaine**.
- Initialize and serialize in the XML handler of **VariablesContainer** is Changed to **UserList**.
- Removed the deprecated logic in the XML handler for **UserVariable**.
- Changed the signature in the **BrickListProtocol** and all classes implementing this protocol.
- Changed `id<UserDataProtocol>` to UserList whenever possible.

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catty/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Verify that the Jira ticket is in the status *Ready for Development*
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [ ] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [ ] Stick to the project’s git workflow (rebase and squash your commits)
- [x] Verify that your changes do not have any conflicts with the base branch
- [ ] After the PR, verify that all CI checks have passed
- [ ] Post a message in the *#catty* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

File: src/CattyTests/Views/BrickCells/Data/BrickCellVariableDataTests.swift [hunk 1]
```diff
@@ -36,7 +36,7 @@ class BrickCellVariableDataTests: XCTestCase {
     var variablesContainer: VariablesContainer!
     var objectVariable1: UserVariable!
     var objectVariable2: UserVariable!
-    var secondObjectVariable: UserVariable!
+    var secondObjectVariable: UserList!
```
Comment (m-herold) [inline src/CattyTests/Views/BrickCells/Data/BrickCellVariableDataTests.swift:39]: Please change `secondObjectVariable` to `UserVariable` (also in the `setUp`).

File: src/CattyTests/Project/ProjectTests.swift [hunk 2]
```diff
@@ -223,14 +223,14 @@ class ProjectsTest: XCTestCase {
         object.name = "newObjectName"
         project.objectList.add(object)
 
-        let list = UserVariable(name: "userList", isList: true)
```
Comment (m-herold) [inline src/CattyTests/Project/ProjectTests.swift:226]: The `testCopyObjectWithObjectList` test should test the copy functionality of a Brick having a local list. Currently, this is tested for a variable, so I would suggest to remove the variable and make use either of a list Brick or use the list in a Formula.

File: src/Catty/DataModel/VariablesContainer/VariablesContainer.h [hunk 3]
```diff
@@ -50,31 +51,31 @@
...
 
 // Array of UserVariable
-- (NSArray*)allVariablesForObject:(SpriteObject*)spriteObject;
-- (NSArray*)allListsForObject:(SpriteObject*)spriteObject;
+- (NSArray<UserVariable*>*)allVariablesForObject:(SpriteObject*)spriteObject;
+- (NSArray<UserList*>*)allListsForObject:(SpriteObject*)spriteObject;
 
 // Array of Variables
-- (NSMutableArray*)allVariables;
+- (NSMutableArray<UserVariable*>*)allVariables;
```
Comment (m-herold) [inline src/Catty/DataModel/VariablesContainer/VariablesContainer.h:67]: Please change to `NSArray<UserVariable*>*`.

File: src/Catty/DataModel/VariablesContainer/VariablesContainer.h [hunk 4]
```diff
@@ -50,31 +51,31 @@
...
 // Array of Lists
-- (NSMutableArray*)allLists;
+- (NSMutableArray<UserList*>*)allLists;
```
Comment (m-herold) [inline src/Catty/DataModel/VariablesContainer/VariablesContainer.h:69]: Please change to `NSArray<UserList*>*`.

File: src/CattyTests/DataModel/VariablesContainerTest.swift [hunk 5]
```diff
@@ -236,32 +233,32 @@ final class VariablesContainerTest: XCTestCase {
...
         let container = VariablesContainer()
 
         container.addObjectList(list1, for: objectA)
-        container.add(toUserList: list1, value: 10)
-        container.add(toUserList: list1, value: 20)
+        container.add(to: list1, value: 10)
+        container.add(to: list1, value: 20)
 
-        let lists = container.allLists() as! [UserVariable]
+        let lists = container.allLists() as! [UserList]
```
Comment (m-herold) [inline src/CattyTests/DataModel/VariablesContainerTest.swift:261]: Please remove the force cast once `NSMutableArray` has been changed to `NSMutableArray` in the `VariablesContainer`.

File: src/Catty/PlayerEngine/Formula/FormulaManager+Interpreter.swift [hunk 6]
```diff
@@ -207,9 +207,11 @@ extension FormulaManager {
...
     private func interpretList(_ formulaElement: FormulaElement, for spriteObject: SpriteObject) -> AnyObject {
         guard let project = spriteObject.project,
-            let list = project.variables.getUserListNamed(formulaElement.value, for: spriteObject),
-            let value = list.value,
-            let listElements = value as? [Any] else { return 0 as AnyObject }
+            let list = project.variables.getUserListNamed(formulaElement.value, for: spriteObject)
+            else { return 0 as AnyObject }
+
+        let value = list.value
+        guard let listElements = value as? [Any] else { return 0 as AnyObject }
```
Comment (m-herold) [inline src/Catty/PlayerEngine/Formula/FormulaManager+Interpreter.swift:214]: Use same guard as above in order to remove duplicated default value logic.

File: src/CattyTests/Bricks/DeleteItemOfUserListBrickTests.swift [hunk 7]
```diff
@@ -74,7 +74,7 @@ final class DeleteItemOfUserListBrickTests: XCTestCase {
             XCTFail("Fatal Error")
         }
 
-        XCTAssertEqual(0, (userList.value as! [AnyObject]).count)
+        XCTAssertEqual(0, (userList.value as [AnyObject]).count)
```
Comment (m-herold) [inline src/CattyTests/Bricks/DeleteItemOfUserListBrickTests.swift:77]: Remove the unnecessary cast.

File: src/CattyTests/Bricks/DeleteItemOfUserListBrickTests.swift [hunk 8]
```diff
@@ -88,7 +88,7 @@ final class DeleteItemOfUserListBrickTests: XCTestCase {
             XCTFail("Fatal Error")
         }
 
-        XCTAssertEqual(1, (userList.value as! [AnyObject]).count)
+        XCTAssertEqual(1, (userList.value as [AnyObject]).count)
```
Comment (m-herold) [inline src/CattyTests/Bricks/DeleteItemOfUserListBrickTests.swift:91]: Remove the unnecessary cast.

File: src/CattyTests/Bricks/DeleteItemOfUserListBrickTests.swift [hunk 9]
```diff
@@ -102,7 +102,7 @@ final class DeleteItemOfUserListBrickTests: XCTestCase {
...
```
Comment (m-herold) [inline src/CattyTests/Bricks/DeleteItemOfUserListBrickTests.swift:105]: Remove the unnecessary cast.

File: src/CattyTests/Bricks/ReplaceItemInUserListBrickTests.swift [hunk 10]
```diff
@@ -77,13 +77,13 @@ final class ReplaceItemInUserListBrickTests: XCTestCase {
             XCTFail("Fatal Error")
         }
 
-        let valueArr = userList.value as! [AnyObject]
+        let valueArr = userList.value as [AnyObject]
```
Comment (m-herold) [inline src/CattyTests/Bricks/ReplaceItemInUserListBrickTests.swift:80]: Remove the unnecessary cast.

File: src/CattyTests/Views/BrickCells/Data/BrickCellVariableDataTests.swift (hunk 1 above)
Comment (DhrubojyotiBis1) [reply src/CattyTests/Views/BrickCells/Data/BrickCellVariableDataTests.swift:39]: Hi @m-herold  `secondObjectVariable` is added as `objectList` using `addObjectList` should I change that to `addObjectVariable`

https://github.com/Catrobat/Catty/blob/13ed308fc66e7b464aefbf42635057a2700c0ca0/src/CattyTests/Views/BrickCells/Data/BrickCellVariableDataTests.swift#L58

Comment (m-herold) [reply src/CattyTests/Views/BrickCells/Data/BrickCellVariableDataTests.swift:39]: Yes, please.

File: src/CattyTests/ViewController/FormulaEditorViewControllerTests.swift [hunk 11]
```diff
@@ -165,14 +165,14 @@ final class FormulaEditorViewControllerTests: XCTestCase {
     }
 
     func testIsListUsedGlobalWithoutBrick() {
-        let list = UserVariable(name: "globalList", isList: true)
+        let list = UserVariable(name: "globalList")
```
Comment (m-herold) [inline src/CattyTests/ViewController/FormulaEditorViewControllerTests.swift:168]: According to the name of the test this should be of type `UserList`.

File: src/CattyTests/UserListFunctionsTest.swift [hunk 12]
```diff
@@ -38,7 +38,7 @@ final class UserListFunctionsTest: XCTestCase {
         let object = SpriteObject()
         object.project = project
 
-        let userVariable = UserVariable(name: "TestList", isList: true)
+        let userVariable = UserVariable(name: "TestList")
```
Comment (m-herold) [inline src/CattyTests/UserListFunctionsTest.swift:41]: According to the name of the test class this should be of type `UserList`.

File: src/CattyTests/PlayerEngine/Functions/List/NumberOfItemsFunctionTest.swift [hunk 13]
```diff
@@ -42,7 +42,7 @@ class NumberOfItemsFunctionTest: XCTestCase {
         XCTAssertEqual(type(of: function).defaultValue, function.value(parameter: -2 as AnyObject), accuracy: Double.epsilon)
         XCTAssertEqual(type(of: function).defaultValue, function.value(parameter: nil), accuracy: Double.epsilon)
 
-        let userVariable = UserVariable(name: "testName", isList: true)
```
Comment (m-herold) [inline src/CattyTests/PlayerEngine/Functions/List/NumberOfItemsFunctionTest.swift:45]: According to the name of the test this should be of type `UserList`.

File: src/CattyTests/PlayerEngine/Functions/List/ElementFunctionTest.swift [hunk 14]
```diff
@@ -43,7 +43,7 @@ class ElementFunctionTest: XCTestCase {
         XCTAssertEqual(type(of: function).defaultValue as! String, function.value(firstParameter: 2 as AnyObject, secondParameter: -3 as AnyObject) as! String)
         XCTAssertEqual(type(of: function).defaultValue as! String, function.value(firstParameter: nil, secondParameter: nil) as! String)
 
-        let userVariableNumber = UserVariable(name: "testName", isList: true)
```
Comment (m-herold) [inline src/CattyTests/PlayerEngine/Functions/List/ElementFunctionTest.swift:46]: According to the name of the test this should be of type `UserList`.

File: src/CattyTests/PlayerEngine/Functions/List/ContainsFunctionTest.swift [hunk 15]
```diff
@@ -43,23 +43,23 @@ class ContainsFunctionTest: XCTestCase {
         XCTAssertEqual(type(of: function).defaultValue, function.value(firstParameter: 100 as AnyObject, secondParameter: 2 as AnyObject), accuracy: Double.epsilon)
         XCTAssertEqual(type(of: function).defaultValue, function.value(firstParameter: nil, secondParameter: nil), accuracy: Double.epsilon)
 
-        let userVariable = UserVariable(name: "testName", isList: true)
```
Comment (m-herold) [inline src/CattyTests/PlayerEngine/Functions/List/ContainsFunctionTest.swift:46]: According to the name of the test this should be of type `UserList`.

File: src/CattyTests/Parser & Serialization/UserListXMLHandlerTests.m [hunk 16]
```diff
@@ -53,7 +53,7 @@ - (void)setUp
     self.serializerContext = [[CBXMLSerializerContext alloc] init];
 
     self.userList = [[UserList alloc] initWithName: @"testUserList"];
-    self.userListOfTypeUserVariable = [[UserVariable alloc] initWithName: @"testUserListOfTypeUserVariable" isList:YES];
+    self.userListOfTypeUserVariable = [[UserVariable alloc] initWithName: @"testUserListOfTypeUserVariable"];
```
Comment (m-herold) [inline src/CattyTests/Parser & Serialization/UserListXMLHandlerTests.m:56]: Please rename to _userVariable_, or remove if not necessary anymore.

File: src/Catty/DataModel/VariablesContainer/VariablesContainer.m [hunk 17]
```diff
@@ -379,7 +394,7 @@ - (NSMutableArray*)allVariables
     return vars;
 }
 
-- (NSMutableArray*)allLists
+- (NSArray*)allLists
```
Comment (m-herold) [inline src/Catty/DataModel/VariablesContainer/VariablesContainer.m:397]: Please add type of array.

File: src/Catty/DataModel/VariablesContainer/VariablesContainer.m [hunk 18]
```diff
@@ -368,7 +383,7 @@ - (NSArray*)allListsForObject:(SpriteObject*)spriteObject
     return lists;
 }
 
-- (NSMutableArray*)allVariables
+- (NSArray*)allVariables
```
Comment (m-herold) [inline src/Catty/DataModel/VariablesContainer/VariablesContainer.m:386]: Please add type of array.

**Pull request merged**